File: crates/osr-interlocking/src/log.rs
//! Entry schema — Rust mirror of `osr-core/proto/track_state.proto`.
//!
//! Integer-only in the safety path. Floats are only used for sensor-input
//! fields that are inherently fractional (speed_mps, speed_uncertainty_mps,
//! pack_state_of_charge); they do not participate in the MA-computation
//! path's authoritative arithmetic.

use osr_core::{
    ConsistDescriptor, EntityId, EntryId, RegionId, RouteId, SectionId, SwitchId,
    TrainId,
};
use osr_core::{Direction, Position, TrackRef};
use serde::{Deserialize, Serialize};

// ---------------------------------------------------------------------------
// Top-level envelope
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Entry {
    pub entry_id: EntryId,
    /// Raft term at which the entry was committed (RFC 0001 §6.4).
    pub term: u64,
    /// Leader's PTP-synced TAI clock, ns since epoch, at commit.
    pub timestamp_ns: u64,
    pub payload: EntryPayload,
}

/// Sum type over all entry variants. The discriminants here match the
/// protobuf `oneof` payload field numbers (10..21).
///
/// `Hash`/`Eq` are not derived because `TrainRegistration` contains a
/// `ConsistDescriptor` with `f32` fields (max_speed_mps, braking curve
/// points). Use `PartialEq` for structural comparison.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub enum EntryPayload {
    TrainPositionReport(TrainPositionReport),
    SwitchCommand(SwitchCommand),
    SwitchObservation(SwitchObservation),
    RouteRequest(RouteRequest),
    RouteGrant(RouteGrant),
    RouteRelease(RouteRelease),
    SpeedRestriction(SpeedRestriction),
    TrainRegistration(TrainRegistration),
    TrainDeparture(TrainDeparture),
    Heartbeat(Heartbeat),
    MaintenanceOverride(MaintenanceOverride),
    FormatVersion(FormatVersion),
}

// ---------------------------------------------------------------------------
// Position report
// ---------------------------------------------------------------------------

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum PositionSource {
    Gnss,
    Imu,
    Odometry,
    Beacon,
    LegacyTrackCircuit,
    LegacyAxleCounter,
}

/// Primary sensor-fusion output, 5 Hz nominal.
/// Speed/uncertainty fields are retained as mm/s i64 values so MA arithmetic
/// stays integer.
#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct TrainPositionReport {
    pub train_id: TrainId,
    pub head_position: Position,
    pub tail_position: Position,
    /// Millimetres per second (positive in Direction::Forward along the line).
    pub speed_mmps: i64,
    /// Half-width 95% CI on speed, in mm/s.
    pub speed_uncertainty_mmps: u32,
    pub heading: Direction,
    pub contributing_sources: Vec<PositionSource>,
    /// TAI ns on the train's own clock.
    pub onboard_time_ns: u64,
    /// 0..=1000 — encoded as parts-per-thousand to stay integer.
    pub pack_soc_ppt: u16,
}

// ---------------------------------------------------------------------------
// Switches
// ---------------------------------------------------------------------------

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum SwitchPosition {
    Normal,
    Reverse,
    Transitioning,
    /// Fail-restrictive default — treat as unsafe until an Observation
    /// says otherwise.
    Unknown,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Confidence {
    Locked,
    Observed,
    Transitioning,
    Unknown,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct SwitchCommand {
    pub switch_id: SwitchId,
    pub requested_position: SwitchPosition,
    pub requested_by: EntityId,
    pub lock_until: Option<EntryId>,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct SwitchObservation {
    pub switch_id: SwitchId,
    pub observed_position: SwitchPosition,
    pub confidence: Confidence,
    pub observed_at_ns: u64,
}

// ---------------------------------------------------------------------------
// Routes
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct RouteRequest {
    pub route_id: RouteId,
    pub requested_by: EntityId,
    pub entry_point: TrackRef,
    pub exit_point: TrackRef,
    pub train_id: Option<TrainId>,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct RouteGrant {
    pub route_id: RouteId,
    pub train_id: TrainId,
    pub locked_switches: Vec<SwitchId>,
    pub locked_sections: Vec<SectionId>,
    pub expires_at_ns: u64,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct RouteRelease {
    pub route_id: RouteId,
    pub reason: String,
}

// ---------------------------------------------------------------------------
// Speed restrictions
// ---------------------------------------------------------------------------

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum RestrictionReason {
    Permanent,
    Temporary,
    Emergency,
    Weather,
    InfrastructureFault,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct SpeedRestriction {
    pub section: SectionId,
    pub from_offset_mm: i64,
    pub to_offset_mm: i64,
    /// Max allowable speed in mm/s.
    pub max_speed_mmps: i64,
    pub reason: RestrictionReason,
    pub effective_from_ns: u64,
    /// `None` means indefinite.
    pub effective_until_ns: Option<u64>,
    pub issued_by: EntityId,
}

// ---------------------------------------------------------------------------
// Train registration / departure
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct TrainRegistration {
    pub train_id: TrainId,
    pub consist: ConsistDescriptor,
    pub initial_position: Position,
    // Cert chain omitted for the in-memory representation; the signed wire
    // form carries it. The signature is verified before an Entry enters
    // the log, not inside derive_state.
}

// `TrainRegistration` contains `ConsistDescriptor`, which has `f32`
// fields (max_speed_mps, braking curve). `PartialEq` composes through to
// `f32` equality — this is imperfect (NaN, -0.0) but acceptable for
// derived-state comparison of well-formed registrations. We intentionally
// do not derive `Eq` or `Hash` for `TrainRegistration` for this reason.

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct TrainDeparture {
    pub train_id: TrainId,
    /// The region this train is handed off to. `None` means leaving the
    /// federated network entirely.
    pub handed_off_to: Option<RegionId>,
    pub handoff_time_ns: u64,
}

// ---------------------------------------------------------------------------
// Liveness & overrides
// ---------------------------------------------------------------------------

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum HealthStatus {
    Ok,
    Degraded,
    Failing,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct Heartbeat {
    pub from_entity: EntityId,
    pub health: HealthStatus,
    pub monotonic_seq: u64,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct MaintenanceOverride {
    pub section: SectionId,
    pub from_offset_mm: i64,
    pub to_offset_mm: i64,
    pub granted_to: EntityId,
    pub granted_until_ns: u64,
    pub rationale: String,
    // Signatures omitted in the derived-state representation for the same
    // reason as TrainRegistration's cert chain.
}

// ---------------------------------------------------------------------------
// Format version
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct FormatVersion {
    pub current: u32,
    pub min_compatible: u32,
    pub schema_sha256_hex: String,
}

File: tools/reference-ma/src/reference_ma/log.py
"""Entry schema — mirror of osr-interlocking::log.

Rust's `EntryPayload` is an externally-tagged enum, so serde produces
`{"TrainRegistration": {...}}`, `{"TrainPositionReport": {...}}`, etc.
We round-trip the same shape here.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any, Dict, List, Optional, Union

from .types import (
    ConsistDescriptor,
    Direction,
    EntityId,
    EntryId,
    Position,
    RegionId,
    RouteId,
    SectionId,
    StationId,
    SwitchId,
    TrackRef,
    TrainId,
)


class PositionSource(str, Enum):
    Gnss = "Gnss"
    Imu = "Imu"
    Odometry = "Odometry"
    Beacon = "Beacon"


class IntrusionState(str, Enum):
    Clear = "Clear"
    Unknown = "Unknown"
    Present = "Present"


class SwitchPosition(str, Enum):
    Normal = "Normal"
    Reverse = "Reverse"
    Transitioning = "Transitioning"
    Unknown = "Unknown"


class Confidence(str, Enum):
    Locked = "Locked"
    Observed = "Observed"
    Transitioning = "Transitioning"
    Unknown = "Unknown"


class HealthStatus(str, Enum):
    Ok = "Ok"
    Degraded = "Degraded"
    Failing = "Failing"


class RestrictionReason(str, Enum):
    Permanent = "Permanent"
    Temporary = "Temporary"
    Emergency = "Emergency"
    Weather = "Weather"
    InfrastructureFault = "InfrastructureFault"


# ---------------------------------------------------------------------------
# Payload variants
# ---------------------------------------------------------------------------


@dataclass
class TrainPositionReport:
    train_id: TrainId
    head_position: Position
    tail_position: Position
    speed_mmps: int
    speed_uncertainty_mmps: int
    heading: Direction
    contributing_sources: List[PositionSource]
    onboard_time_ns: int
    pack_soc_ppt: int

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "TrainPositionReport":
        return cls(
            train_id=d["train_id"],
            head_position=Position.from_json(d["head_position"]),
            tail_position=Position.from_json(d["tail_position"]),
            speed_mmps=int(d["speed_mmps"]),
            speed_uncertainty_mmps=int(d["speed_uncertainty_mmps"]),
            heading=Direction(d["heading"]),
            contributing_sources=[PositionSource(s) for s in d["contributing_sources"]],
            onboard_time_ns=int(d["onboard_time_ns"]),
            pack_soc_ppt=int(d["pack_soc_ppt"]),
        )


@dataclass
class SwitchCommand:
    switch_id: SwitchId
    requested_position: SwitchPosition
    requested_by: EntityId
    lock_until: Optional[EntryId]

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "SwitchCommand":
        return cls(
            switch_id=d["switch_id"],
            requested_position=SwitchPosition(d["requested_position"]),
            requested_by=d["requested_by"],
            lock_until=d.get("lock_until"),
        )


@dataclass
class SwitchObservation:
    switch_id: SwitchId
    observed_position: SwitchPosition
    confidence: Confidence
    observed_at_ns: int

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "SwitchObservation":
        return cls(
            switch_id=d["switch_id"],
            observed_position=SwitchPosition(d["observed_position"]),
            confidence=Confidence(d["confidence"]),
            observed_at_ns=int(d["observed_at_ns"]),
        )


@dataclass
class RouteRequest:
    route_id: RouteId
    requested_by: EntityId
    entry_point: TrackRef
    exit_point: TrackRef
    train_id: Optional[TrainId]

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "RouteRequest":
        return cls(
            route_id=d["route_id"],
            requested_by=d["requested_by"],
            entry_point=TrackRef.from_json(d["entry_point"]),
            exit_point=TrackRef.from_json(d["exit_point"]),
            train_id=d.get("train_id"),
        )


@dataclass
class RouteGrant:
    route_id: RouteId
    train_id: TrainId
    locked_switches: List[SwitchId]
    locked_sections: List[SectionId]
    expires_at_ns: int

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "RouteGrant":
        return cls(
            route_id=d["route_id"],
            train_id=d["train_id"],
            locked_switches=list(d["locked_switches"]),
            locked_sections=list(d["locked_sections"]),
            expires_at_ns=int(d["expires_at_ns"]),
        )


@dataclass
class RouteRelease:
    route_id: RouteId
    reason: str

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "RouteRelease":
        return cls(route_id=d["route_id"], reason=d["reason"])


@dataclass
class SpeedRestriction:
    section: SectionId
    from_offset_mm: int
    to_offset_mm: int
    max_speed_mmps: int
    reason: RestrictionReason
    effective_from_ns: int
    effective_until_ns: Optional[int]
    issued_by: EntityId

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "SpeedRestriction":
        return cls(
            section=d["section"],
            from_offset_mm=int(d["from_offset_mm"]),
            to_offset_mm=int(d["to_offset_mm"]),
            max_speed_mmps=int(d["max_speed_mmps"]),
            reason=RestrictionReason(d["reason"]),
            effective_from_ns=int(d["effective_from_ns"]),
            effective_until_ns=(
                int(d["effective_until_ns"])
                if d.get("effective_until_ns") is not None
                else None
            ),
            issued_by=d["issued_by"],
        )


@dataclass
class SectionIntrusion:
    section: SectionId
    state: IntrusionState
    issued_by: EntityId
    observed_at_ns: int

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "SectionIntrusion":
        return cls(
            section=d["section"],
            state=IntrusionState(d["state"]),
            issued_by=d["issued_by"],
            observed_at_ns=int(d["observed_at_ns"]),
        )


@dataclass
class TrainRegistration:
    train_id: TrainId
    consist: ConsistDescriptor
    initial_position: Position

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "TrainRegistration":
        return cls(
            train_id=d["train_id"],
            consist=ConsistDescriptor.from_json(d["consist"]),
            initial_position=Position.from_json(d["initial_position"]),
        )


@dataclass
class TrainDeparture:
    train_id: TrainId
    handed_off_to: Optional[RegionId]
    handoff_time_ns: int

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "TrainDeparture":
        return cls(
            train_id=d["train_id"],
            handed_off_to=d.get("handed_off_to"),
            handoff_time_ns=int(d["handoff_time_ns"]),
        )


@dataclass
class Heartbeat:
    from_entity: EntityId
    health: HealthStatus
    monotonic_seq: int

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "Heartbeat":
        return cls(
            from_entity=d["from_entity"],
            health=HealthStatus(d["health"]),
            monotonic_seq=int(d["monotonic_seq"]),
        )


@dataclass
class MaintenanceOverride:
    section: SectionId
    from_offset_mm: int
    to_offset_mm: int
    granted_to: EntityId
    granted_until_ns: int
    rationale: str

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "MaintenanceOverride":
        return cls(
            section=d["section"],
            from_offset_mm=int(d["from_offset_mm"]),
            to_offset_mm=int(d["to_offset_mm"]),
            granted_to=d["granted_to"],
            granted_until_ns=int(d["granted_until_ns"]),
            rationale=d["rationale"],
        )


@dataclass
class FormatVersion:
    current: int
    min_compatible: int
    schema_sha256_hex: str

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "FormatVersion":
        return cls(
            current=int(d["current"]),
            min_compatible=int(d["min_compatible"]),
            schema_sha256_hex=d["schema_sha256_hex"],
        )


# Union type for an EntryPayload. Kept as a plain tuple `(tag, value)`
# after decoding — one variant per tag.
EntryPayload = Union[
    TrainPositionReport,
    SwitchCommand,
    SwitchObservation,
    RouteRequest,
    RouteGrant,
    RouteRelease,
    SpeedRestriction,
    SectionIntrusion,
    TrainRegistration,
    TrainDeparture,
    Heartbeat,
    MaintenanceOverride,
    FormatVersion,
]


_PAYLOAD_DECODERS = {
    "TrainPositionReport": TrainPositionReport.from_json,
    "SwitchCommand": SwitchCommand.from_json,
    "SwitchObservation": SwitchObservation.from_json,
    "RouteRequest": RouteRequest.from_json,
    "RouteGrant": RouteGrant.from_json,
    "RouteRelease": RouteRelease.from_json,
    "SpeedRestriction": SpeedRestriction.from_json,
    "SectionIntrusion": SectionIntrusion.from_json,
    "TrainRegistration": TrainRegistration.from_json,
    "TrainDeparture": TrainDeparture.from_json,
    "Heartbeat": Heartbeat.from_json,
    "MaintenanceOverride": MaintenanceOverride.from_json,
    "FormatVersion": FormatVersion.from_json,
}


def _decode_payload(obj: Dict[str, Any]) -> EntryPayload:
    if len(obj) != 1:
        raise ValueError(
            f"expected externally-tagged enum with one key, got {list(obj)}"
        )
    tag, inner = next(iter(obj.items()))
    decoder = _PAYLOAD_DECODERS.get(tag)
    if decoder is None:
        raise ValueError(f"unknown EntryPayload variant {tag!r}")
    return decoder(inner)


@dataclass
class Entry:
    entry_id: EntryId
    term: int
    timestamp_ns: int
    payload: EntryPayload

    @classmethod
    def from_json(cls, d: Dict[str, Any]) -> "Entry":
        return cls(
            entry_id=d["entry_id"],
            term=int(d["term"]),
            timestamp_ns=int(d["timestamp_ns"]),
            payload=_decode_payload(d["payload"]),
        )
